File: leetcode/must-do-medium-questions/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) return 0;

        int num = 0;
        vector<vector<int>> visited(grid.size(), vector<int>(grid[0].size(), 0));
        for(int row = 0; row < grid.size(); row++) {
            for(int col = 0; col < grid[0].size(); col++) {
                queue<pair<int, int>> q;
                if (grid[row][col] == '1' && !visited[row][col]) {
                    q.push(make_pair(row, col));
                    visited[row][col] = 1;
                    num++;
                }
                while (!q.empty()) {
                    pair<int, int> point = q.front();
                    q.pop();
                    int r = point.first;
                    int c = point.second;
                    if (r-1 >= 0 && !visited[r-1][c] && grid[r-1][c] == '1') {
                        q.push(make_pair(r-1, c));
                        visited[r-1][c] = 1;
                    }
                    if (c-1 >= 0 && !visited[r][c-1] && grid[r][c-1] == '1') {
                        q.push(make_pair(r, c-1));
                        visited[r][c-1] = 1;
                    }
                    if (r+1 < grid.size() && !visited[r+1][c] && grid[r+1][c] == '1') {
                        q.push(make_pair(r+1, c));
                        visited[r+1][c] = 1;
                    }
                    if (c+1 < grid[0].size() && !visited[r][c+1] && grid[r][c+1] == '1') {
                        q.push(make_pair(r, c+1));
                        visited[r][c+1] = 1;
                    }
                }
            }
        }
        return num;
    }
};

File: leetcode/30-day-leetcoding-challenge/number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
const numIslands = function(grid) {
    if (grid.length == 0) return 0;
    
    let cnt = 0;
    for(let row = 0; row < grid.length; row++) {
        for(let col = 0; col < grid[0].length; col++) {
            if (grid[row][col] == 1) {
                const q = [];
                grid[row][col] = 0;
                q.push([row,col]);
                while(q.length > 0) {
                    const [r, c] = q.shift();
                    if (r-1 >= 0 && grid[r-1][c] == 1) {
                        grid[r-1][c] = 0;
                        q.push([r-1, c]);
                    }
                    if (c-1 >= 0 && grid[r][c-1] == 1) {
                        grid[r][c-1] = 0;
                        q.push([r, c-1]);
                    }
                    if (r+1 < grid.length && grid[r+1][c] == 1) {
                        grid[r+1][c] = 0;
                        q.push([r+1, c]);
                    }
                    if (c+1 < grid[0].length && grid[r][c+1] == 1) {
                        grid[r][c+1] = 0;
                        q.push([r, c+1]);
                    }
                }
                cnt += 1;
            }
        }
    }
    return cnt;
};


